src/lib/optimizer/chunk_statistics.hpp

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include <exception>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include "all_type_variant.hpp"
+#include "type_cast.hpp"
+#include "types.hpp"
+
+#include "storage/base_encoded_column.hpp"
+#include "storage/deprecated_dictionary_column.hpp"
+#include "storage/dictionary_column.hpp"
+#include "storage/reference_column.hpp"
+#include "storage/run_length_column.hpp"
+#include "storage/value_column.hpp"
+
+namespace opossum {
+
+class BaseFilter : public std::enable_shared_from_this<BaseFilter> {
+ public:
+  virtual ~BaseFilter() = default;
+
+  virtual bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const = 0;
+};
+
+class ChunkColumnStatistics {
+ public:
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics(DataType data_type, std::shared_ptr<BaseColumn> column);
+
+  void add_filter(std::shared_ptr<BaseFilter> filter) {
+    _filters.emplace_back(filter);
+  }
+
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const {
+    for(const auto& filter : _filters) {
+      if(filter->can_prune(value, predicate_type)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+ protected:
+  template<typename T>
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_dictionary(const pmr_vector<T>& dictionary);
+
+  template<typename T>
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const DictionaryColumn<T>& column) {
+    const auto & dictionary = *column.dictionary();
+    return build_statistics_from_dictionary(dictionary);
+  }
+
+  template<typename T>
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const ValueColumn<T>& column) {
+  DebugAssert(false, "Chunk statistics should only be computed for compressed columns!");
+  return std::make_shared<ChunkColumnStatistics>();
+  }
+
+  template<typename T>
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const DeprecatedDictionaryColumn<T>& column) {
+    const auto & dictionary = *column.dictionary();
+    return build_statistics_from_dictionary(dictionary);
+  }
+
+  template<typename T>
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const ReferenceColumn& column) {
+  DebugAssert(false, "Chunk statistics should only be computed for compressed columns!");
+  return std::make_shared<ChunkColumnStatistics>();
+  }
+
+  template<typename T>
+  static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const RunLengthColumn<T>& column) {
+  DebugAssert(false, "Not Implemented!");
+  return std::make_shared<ChunkColumnStatistics>();
```

Comment:
Please try this instead:
```c++
template<typename T>
static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const DictionaryColumn<T>& column) {
  const auto& dictionary = *column.dictionary();
  return build_statistics_from_dictionary(dictionary);
}

template<typename T>
static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const DeprecatedDictionaryColumn<T>& column) {
  const auto& dictionary = *column.dictionary();
  return build_statistics_from_dictionary(dictionary);
}

static std::shared_ptr<ChunkColumnStatistics> build_statistics_from_concrete_column(const BaseColumn& column) {
  Fail("Your nice error message");
}
```

Reply:
Does not work unfortunately. I think it has something to do with the way `build_statistics` is implemented, where we call `resolve_data_and_column_type`.

Reply:
This worked for me: https://gist.github.com/mjendruk/01c9eb3a65a539584f10aae6a33954ee

Reply:
@tbjoern Yes it does, there was just a tiny change needed in the calling code. I checked out your branch and compiled it myself. If you haven’t changed the `chunk_statistics.hpp` the code in the linked file should still work. Just replace the entire file with the one from the link.